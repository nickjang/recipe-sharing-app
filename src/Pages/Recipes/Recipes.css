.recipes-page-title {
  height: 94px;
}