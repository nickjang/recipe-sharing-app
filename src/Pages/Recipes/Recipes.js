import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import RecipeSharingApiService from '../../services/recipe-sharing-api-service';
import InfiniteScroll from 'react-infinite-scroll-component';
import './Recipes.css';

class Recipes extends Component {
  state = {
    recipes: [],
    hasMore: true,
    offset: 0, // number recipes already fetched
    error: ''
  }

  fetchRecipes = async () => {
    this.setState({ error: '' });
    try {
      let offset = this.state.offset;
      // fetch
      let { recipes, hasMore } = await RecipeSharingApiService.getMoreRecipes(10, offset);
      // increase offset
      offset += recipes.length;
      // add new recipes to end
      recipes = this.state.recipes.concat(recipes)
      this.setState({ recipes, hasMore, offset });
    } catch ({ error }) {
      this.setState({ error });
    }
  }

  componentDidMount = async () => {
    this.setState({ error: '' });
    try {
      await this.fetchRecipes();
    } catch ({ error }) {
      this.setState({ error });
    }
  }

  generateRecipeCards = (recipes) => {
    return recipes.map((recipe, index) => {
      let timestamp;
      const dateObj = new Date(recipe.date_modified || recipe.date_created);
      if (recipe.date_modified) {
        timestamp = 'Updated: ';
        timestamp += dateObj.toDateString();
      } else {
        timestamp = 'Published: ';
        timestamp += dateObj.toDateString();
      }
      return <article key={`${index}-recipe`} className='rs-card recipe-card'>
        <Link
          to={`/recipes/${recipe.id}`}
          className='rs-sub-title'
        >
          {recipe.name}
        </Link>
        <p className='note'>{`Author: ${recipe.author.full_name}`}</p>
        <p className='note recipe-information'>{recipe.information}</p>
        <p className='timestamp'>{timestamp}</p>
      </article>
    });
  }

  generateGroupedCards = (recipes) => {
    recipes = this.generateRecipeCards(recipes);
    let size = 3;
    let recipeIdx = 0;
    let groupIdx = -1;
    let length = recipes.length;
    const groups = [];
    while (recipeIdx < length) {
      if (recipeIdx % size === 0) {
        groupIdx++;
        groups[groupIdx] = [];
      }
      groups[groupIdx].push(recipes[recipeIdx]);
      recipeIdx++;
    }

    // return the grouped cards
    return groups.map(group =>
      <div key={`${group}-group`} className='recipe-group'>
        {group}
      </div>
    );
  }

  render() {
    const { recipes, hasMore, error } = this.state;

    return (
      <InfiniteScroll
        dataLength={recipes.length}
        next={this.fetchRecipes}
        hasMore={hasMore}
        loader={<span>Loading...</span>}
        endMessage={error
          ? <p className='error text-center'>{error}</p>
          : <p className='note text-center'>End of recipes</p>}
        className='recipe-card-list'
      > {this.generateGroupedCards(recipes)}
      </InfiniteScroll >
    );
  }
}

export default Recipes;