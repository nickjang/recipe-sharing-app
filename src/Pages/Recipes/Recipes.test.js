import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Recipes from './Recipes';

describe(`Recipes component`, () => {
  it('renders Recipes by default', () => {
    const wrapper = shallow(<Recipes />);
    expect(toJson(wrapper)).toMatchSnapshot()
  });
})