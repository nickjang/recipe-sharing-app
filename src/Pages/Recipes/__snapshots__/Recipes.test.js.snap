// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recipes component renders Recipes by default 1`] = `
<article>
  <Parallax
    bgClassName="react-parallax-bgimage"
    bgImage="cupcakes.jpg"
    bgImageAlt="Chocolate cupcakes stacked with chocolate powder sprinkling on top"
    blur={0}
    className="recipes-page-title"
    contentClassName=""
    disabled={false}
    strength={200}
  />
  <InfiniteScroll
    className="recipe-card-list"
    dataLength={0}
    endMessage={
      <p
        className="note text-center"
      >
        End of recipes
      </p>
    }
    hasMore={true}
    loader={
      <span>
        Loading...
      </span>
    }
    next={[Function]}
  >
     
  </InfiniteScroll>
</article>
`;
